### PR TITLE
Convert QB to typescript

### PR DESCRIPTION
**Pull request by oisincoveney**
I look at this component a lot and I realized I could do this on a free night. The steps were:
- Change the file name
- Add props type for redux
- Add modal and modalContext keys to `QueryBuilderUIControls`
- Added types for parameters and loaders

### CODE REVIEW

File: frontend/src/metabase/query_builder/containers/QueryBuilder.tsx [hunk 1]
```diff
@@ -94,96 +101,120 @@ const timelineProps = {
...
+
+type QueryBuilderInnerProps = ReduxProps &
+  WithRouterProps &
+  BookmarkLoadListProps &
+  TimelineLoadListProps & {
+    route: Route;
+  };
+
+function QueryBuilderInner(props: QueryBuilderInnerProps) {
+  const isBookmarked = props.bookmarks.some(
```
Comment (npretto) [inline frontend/src/metabase/query_builder/containers/QueryBuilder.tsx:214]: Why was this inlined? Didn't it use a selector before?
It seems like we have 2+ implementation of this logic or similar

Comment (oisincoveney) [reply frontend/src/metabase/query_builder/containers/QueryBuilder.tsx:214]: When I was trying to get the types to work, I was losing the types when we use the `props` in `mapStateToProps`

Comment (oisincoveney) [reply frontend/src/metabase/query_builder/containers/QueryBuilder.tsx]: I had another go at things and it looks like I was able to get the types to work once I added some types for the loaders. I think it should be good to go now

File: frontend/src/metabase-types/store/qb.ts [hunk 2]
```diff
@@ -37,6 +39,8 @@ export interface QueryBuilderUIControls {
   isShowingNotebookNativePreview: boolean;
   notebookNativePreviewSidebarWidth: number | null;
   showSidebarTitle: boolean;
+  modal: QueryModalType | null;
+  modalContext: TimelineEventId | null;
```
Comment (kamilmielnik) [inline frontend/src/metabase-types/store/qb.ts:43]: These attributes are currently undefined in `DEFAULT_UI_CONTROLS` (initial state of the reducer), let's add them there.

File: frontend/src/metabase/query_builder/containers/QueryBuilder.tsx [hunk 3]
```diff
@@ -94,11 +102,30 @@ const timelineProps = {
...
+type TimelineListLoaderOutput = {
+  timelines: Timeline[];
+  reloadTimelines: () => void;
+};
+
+type EntityListLoaderMergedProps = {
+  allLoading: boolean;
+  allLoaded: boolean;
+  allFetched: boolean;
+  allError: boolean;
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/containers/QueryBuilder.tsx:119]:
```suggestion
  allError: unknown;
```

File: frontend/src/metabase/query_builder/containers/QueryBuilder.tsx [hunk 4]
```diff
@@ -213,7 +251,7 @@ function QueryBuilder(props) {
     () => _.debounce(forceUpdate, 400),
     [forceUpdate],
   );
-  const timeout = useRef(null);
+  const timeout = useRef<NodeJS.Timeout>();
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/containers/QueryBuilder.tsx:254]: Looks like nodejs typings are interfering with dom typings.
But it's `number` in the browser: https://github.com/microsoft/TypeScript/blob/2c23bea/src/lib/dom.generated.d.ts#L28698

```suggestion
  const timeout = useRef<number>();
```

**Review by kamilmielnik — COMMENTED** [review]
Nice :+1:

File: frontend/src/metabase-types/store/qb.ts (hunk 2 above)
Comment (oisincoveney) [reply frontend/src/metabase-types/store/qb.ts:43]: Way ahead of you 😄 https://github.com/metabase/metabase/pull/49276/files

Comment (kamilmielnik) [reply frontend/src/metabase-types/store/qb.ts:43]: weird, I don't see it there :)

Comment (oisincoveney) [reply frontend/src/metabase-types/store/qb.ts:43]: I thought I could change the branch faster 😂 should be there now

File: frontend/src/metabase/query_builder/containers/QueryBuilder.tsx (hunk 4 above)
Comment (oisincoveney) [reply frontend/src/metabase/query_builder/containers/QueryBuilder.tsx]: I followed the suggestion but it seems like at 
```
timeout.current = setTimeout(forceUpdateDebounced, 300);
```

we get the whole error of `number` not assignable to `NodeJS.Timeout`. I looked at this [StackOverflow thread](https://stackoverflow.com/a/60247753) and we have two options:

- We _could_ use `window.setTimeout` which will use the TS type that you've linked to. 
- use `useRef<ReturnType<typeof setTimeout>>()` but that still seems to use the `NodeJS.Timeout` under the hood

File: frontend/src/metabase/query_builder/containers/QueryBuilder.tsx (hunk 3 above)
Comment (kamilmielnik) [reply frontend/src/metabase/query_builder/containers/QueryBuilder.tsx:119]: [Trust me, I'm an engineer](https://www.youtube.com/watch?v=rp8hvyjZWHs).

----

1. Apply this patch:

```patch
diff --git a/frontend/src/metabase/entities/containers/EntityListLoader.jsx b/frontend/src/metabase/entities/containers/EntityListLoader.jsx
index 10c7b54ae5..eee1a31259 100644
--- a/frontend/src/metabase/entities/containers/EntityListLoader.jsx
+++ b/frontend/src/metabase/entities/containers/EntityListLoader.jsx
@@ -277,6 +277,8 @@ const EntityListLoader = _.compose(
       entityQuery,
     });
 
+    console.log(error || (allError == null ? null : allError));
+
     return {
       list,
       entityQuery,
```

2. Go to http://localhost:3000/browse/databases/1-sample-database
3. Block network request to GET http://localhost:3000/api/database/1/schemas
4. Refresh the page
5. Be amazed due to lack of booleans in JS console

File: frontend/src/metabase/query_builder/containers/QueryBuilder.tsx (hunk 4 above)
Comment (kamilmielnik) [reply frontend/src/metabase/query_builder/containers/QueryBuilder.tsx]: It seems that we're using almost all possible approaches in the codebase:
- `any` https://github.com/metabase/metabase/blob/92d615073a3ba31710e26be9f2e83141352d1d1c/frontend/src/metabase/nav/containers/MainNavbar/SidebarItems/SidebarCollectionLink.tsx#L48
- `NodeJS.Timeout` https://github.com/metabase/metabase/blob/d897f2016fa86e88e1e4936a53ef14a64c803db6/frontend/src/metabase/hooks/use-action-button-label/use-action-button-label.ts#L18
- `ReturnType<typeof setTimeout>` https://github.com/metabase/metabase/blob/92d615073a3ba31710e26be9f2e83141352d1d1c/frontend/src/metabase/ui/components/utils/DelayGroup/DelayGroup.tsx#L35

Now we just need `window.setTimeout` to get the completionist achievement.

---

The only one that gives proper typing is `window.setTimeout`.

File: frontend/src/metabase/query_builder/containers/QueryBuilder.tsx (hunk 3 above)
Comment (oisincoveney) [reply frontend/src/metabase/query_builder/containers/QueryBuilder.tsx:119]: you do tend to bring wonder and amazement into my engineering career 😄

**Review by kamilmielnik — APPROVED** [review]
LGTM :+1: 

No [brownie points](https://github.com/metabase/metabase/pull/49035#discussion_r1820604509) though.

File: frontend/src/metabase/query_builder/containers/QueryBuilder.tsx (hunk 3 above)
Comment (oisincoveney) [reply frontend/src/metabase/query_builder/containers/QueryBuilder.tsx:119]: certainly doesn't look like a boolean 😄 
image

**Comment by oisincoveney** [comment]
> LGTM 👍
> 
> No [brownie points](https://github.com/metabase/metabase/pull/49035#discussion_r1820604509) though.

I'm working on it! I want some brownie points 😄

**Pull request merged**